File: src/ParticleSwarmOptimizer.m
%% Implementation of the Particle Swarm Optimization Algorithm
%% Author: Pietro Mosca
%% Email: pietromosca1994@gmail.com
%% Date: 04.02.2021
%% Reference: Kennedy, J.; Eberhart, R. (1995). "Particle Swarm Optimization". Proceedings of IEEE International Conference on Neural Networks. IV. pp. 1942–1948.

% Arguments
% fun:     function handle
% domain   struct(fields: hi, lo)   search domain
% swarm_param   struct(fields: n_particles, dimensions, sampling_method)
% alg_param     struct(fields: CR, F)
% verbose 	 	verbosity mode switch
% log_active 	log active flag

function [gbest_x, gbest_y, log]=ParticleSwarmOptimizer(fun, swarm, topology, domain, alg_param, verbose, log_active) 
    i=1;
    
    % Algorithm
    while i<alg_param.n_iter

        topology.update_position(swarm, alg_param, fun);
        topology.update_velocity(swarm, alg_param);
        topology.update_gbest(swarm);
        topology.update_pbest(swarm);
        
        % verbose
        if verbose>0
            if rem(i, verbose)==0
                figure('Name', ['Step', num2str(i)]);
                swarm.plot(fun, domain);
            end
        end
        
        % log 
        if log_active==true
            log.gbest_x(i, :)=swarm.gbest_x;
            log.gbest_y(i)=swarm.gbest_y;
            
        else
            log.gbest_x=[];
            log.gbest_y=[];
        end
    
        i=i+1;
    end
    
    gbest_x=swarm.gbest_x;
    gbest_y=swarm.gbest_y;
    
end

File: src/cpp/src/ParticleSwarmOptimizer.cpp
#include "ParticleSwarmOptimizer.h"

vector<double> ParticleSwarmOptimizer(double (*func)(vector<double>), int _n_particles, domain_struct _x_domain, alg_param_struct _alg_param){
    // swarm initialization
    swarm _swarm(_n_particles, _x_domain, func);

    // topology initialization
    topology _topology(1);

    vector<double> gbest_x(_swarm.n_dimensions, 0.0);
    
    for (int i=0; i<_alg_param.n_iter; i++){
        _topology.update_position(_swarm, _alg_param, func);
        _topology.update_velocity(_swarm, _alg_param);
        _topology.update_gbest(_swarm);\
        _topology.update_pbest(_swarm);
    };

    gbest_x=_swarm.gbest_x;

    return gbest_x;

}
